### PR TITLE
[Dev][feat] Support A2A Overlap for Megatron-FSDP

### PR DESCRIPTION
**Pull request by Wohox**
# What does this PR do ?
PR for main: https://github.com/NVIDIA/Megatron-LM/pull/3797
Based on: https://github.com/NVIDIA/Megatron-LM/pull/3766 

### Supported Feature Combinations
```
# without param sharding
--use-megatron-fsdp --data-parallel-sharding-strategy "optim_grads" --overlap-moe-expert-parallel-comm
--use-megatron-fsdp --data-parallel-sharding-strategy "optim_grads" --overlap-moe-expert-parallel-comm --delay-wgrad-compute


# param sharding
--use-megatron-fsdp --data-parallel-sharding-strategy "optim_grads_params" --overlap-moe-expert-parallel-comm
--use-megatron-fsdp --data-parallel-sharding-strategy "optim_grads_params" --overlap-moe-expert-parallel-comm --delay-wgrad-compute
```

### Problem

The EP overlap schedule (`overlap_moe_expert_parallel_comm`) interleaves forward and backward passes at the sub-module level (attn, mlp, moe_dispatch, moe_combine) to overlap Expert Parallel A2A communication with compute. This schedule bypasses `TransformerLayer.forward()` entirely, calling sub-modules directly. As a result, FSDP hooks registered on `TransformerLayer` — including parameter all-gather, parameter release, gradient reduction, and root pre/post backward — are never triggered, making FSDP with `optim_grads_params` incompatible with this schedule.

### FSDP Hook Triggering: Standard vs 1F1B Overlap

| Hook | Standard Path (non-1F1B) | 1F1B Overlap Path | Reason for Difference |
|---|---|---|---|
| **`_replace_param_with_raw_if_needed`** | Called inside `MegatronFSDP.forward()` at the start of every forward call. | Called **once** at the top of `combined_1f1b_schedule_for_no_pipelining`, before any schedule phase begins. | The overlap schedule bypasses `MegatronFSDP.forward()` entirely — it calls `GPTModel.build_schedule_plan()` on the unwrapped module directly. The param swap must happen explicitly before the schedule accesses layers. |
| **`_pre_forward_param_unshard`** (per-layer all-gather) | Registered as `nn.Module.register_forward_pre_hook` on each FSDP unit. Triggered **automatically** by PyTorch when `TransformerLayer.forward()` is called. | **Same mechanism** — still triggered by PyTorch forward pre-hooks. With `enable_fine_grained_param_gather=True`, hooks are registered on **every sub-module** (not just FSDP units), giving finer-grained all-gather. | The schedule still calls each layer's `forward()` via `ScheduleNode`, which triggers the registered forward pre-hooks normally. No manual intervention needed. |
| **`_post_forward` (post_forward_release_module)** — per-layer param release after forward | Registered as `nn.Module.register_forward_hook` on each FSDP unit. Triggered **automatically** by PyTorch after `TransformerLayer.forward()` returns. | Called **manually** by the schedule plan via `set_fsdp_reshard_hooks`. Wired in `combined_forward_backward_step` onto each `TransformerLayerSchedulePlan`, and invoked after the last forward node (mlp/moe_combine) of each layer completes. | The schedule decomposes `TransformerLayer.forward()` into sub-ops (attn, mlp, moe_dispatch, moe_combine). The registered `register_forward_hook` on `TransformerLayer` is **never fired** because the schedule never calls `TransformerLayer.forward()` as a whole — it calls sub-modules directly. So the release must be done explicitly. |
| **`_post_backward_release_module`** — per-layer param release after backward | Registered via `_register_post_backward_hook` (a `register_forward_pre_hook` that inserts a `RegisterFSDPBackwardFunction` into the autograd graph). Triggered **automatically** by autograd during backward. | Called **manually** by the schedule plan via `set_fsdp_reshard_hooks`. Wired onto `TransformerLayerSchedulePlan.attn` (the last backward node of each layer). | Same reason as `_post_forward`: the schedule runs backward per-sub-op, not per-`TransformerLayer`. The autograd-based post-backward hook on the FSDP unit output is never reached because the layer-level forward was never called as a whole. |
| **`_pre_backward_param_unshard`** (per-layer backward all-gather) | Attached via `create_custom_backward_hook` (a `register_forward_hook` that registers `register_multi_grad_hook` on the output tensors). Triggered **automatically** by autograd when gradients arrive at the layer's output. | With `enable_fine_grained_param_gather=True`, hooks are registered on **every sub-module** via `create_custom_backward_hook`. Triggered **automatically** by autograd when gradients flow through each sub-module. | When `enable_fine_grained_param_gather=False`, hooks only exist on FSDP units; since the layer-level forward is skipped, those hooks are never hit. With `=True`, hooks exist on every sub-module, so autograd still triggers them during the schedule's backward. |
| **`_root_pre_backward`** | Attached via `create_custom_backward_hook` on the root module's output. Triggered **automatically** by autograd (gradient flows back through root output). Also queues `_root_post_backward` via `queue_callback`. | Called **manually** via `fsdp_wrapper.pre_backward()` at the start of each `combined_forward_backward_step` that has a backward model (`b_model is not None`). Called with `skip_backward_hook=True` so it does **not** queue `_root_post_backward`. | The overlap schedule does not call `MegatronFSDP.forward()`, so no root-level output tensor exists in the autograd graph to trigger the hook. It must be called manually. `skip_backward_hook=True` because `_root_post_backward` is also called manually. |
| **`_root_post_backward`** (grad accumulation + reduce-scatter) | Queued via `torch.autograd.Variable._execution_engine.queue_callback()` inside `_root_pre_backward`. Triggered **automatically** by the autograd engine after the entire backward pass finishes. | Called **manually** via `fsdp_wrapper.post_backward()` at the end of each `combined_forward_backward_step` that has a backward model. | Since `_root_pre_backward` is called with `skip_backward_hook=True`, the autograd `queue_callback` is skipped. The schedule must call `post_backward()` explicitly after all layer backward ops complete to trigger gradient accumulation and reduce-scatter. |
| **`_process_post_backward_gradients`** (per-param grad accumulation) | Registered via `param.register_post_accumulate_grad_hook`. Triggered **automatically** by autograd when `.grad` is accumulated on each parameter. | **Same mechanism** w.o. `delay_wgrad_compute` — still triggered automatically by autograd. Parameters are the same leaf tensors regardless of which code path invokes the computation. **Manual Triggered** with `delay_wgrad_compute` since grad acc hook needs to be delayed as well. | This is a parameter-level hook (not module-level), so it fires whenever autograd writes to `.grad`, independent of whether the compute was driven by `module.forward()` or the schedule. |

**Summary**: The root cause of all differences is that the 1F1B overlap schedule **decomposes** `TransformerLayer.forward()` into sub-operations (attn, mlp, moe_dispatch, moe_combine) and interleaves them across microbatches. This means:

1. `MegatronFSDP.forward()` is never called → `_replace_param_with_raw_if_needed` must be explicit.
2. `TransformerLayer.forward()` is never called as a whole → module-level `register_forward_hook` / `register_forward_pre_hook` on the FSDP unit are not fired → `_post_forward` and `_post_backward_release_module` must be wired manually via `set_fsdp_reshard_hooks`.
3. No root-level output tensor in autograd graph → `_root_pre_backward` / `_root_post_backward` must be called manually.
4. Hooks that operate at the **parameter level** (`_process_post_backward_gradients`) or **sub-module level** (`_pre_forward_param_unshard`, `_pre_backward_param_unshard` with `enable_fine_grained=True`) continue to work automatically because the schedule still calls sub-module forwards, and autograd still writes to parameter `.grad`.

### Solution

Bridge the gap between Megatron-FSDP's hook-based lifecycle and the EP overlap schedule's direct sub-module invocation by:
1. Enabling fine-grained per-sub-module all-gather hooks so each sub-module can unshard its own parameters.
2. Exposing manual handles for FSDP lifecycle events that the schedule must drive explicitly.

### Code changes

**`combined_1f1b.py`** — Schedule-level FSDP integration
- Add `find_megatron_fsdp()` helper to locate the FSDP wrapper from the model chain.
- Call `fsdp_wrapper._replace_param_with_raw_if_needed()` before the schedule starts (normally done inside `MegatronFSDP.forward()`).
- Manually invoke `fsdp_wrapper.pre_backward()` / `post_backward()` around the combined forward+backward step since the schedule bypasses the autograd hooks that normally trigger them.
- Wire per-layer FSDP release callbacks via `set_fsdp_reshard_hooks()`.

**`megatron_fsdp.py`** — Expose manual hook handles & support delayed wgrad
- Add `skip_backward_hook` parameter to `_root_pre_backward` so the manual caller can prevent the redundant `queue_callback(_root_post_backward)`.
- Expose `pre_backward`, `post_backward`, `post_forward_release_module`, and `post_backward_release_module` as public attributes for external schedule use.
- Add `is_delayed` parameter to `_process_post_backward_gradients` to distinguish delayed wgrad gradient processing from the standard path, preventing deferred parameters from being filtered out by the `skip_backward_post_hook` guard.
- Register fine-grained backward all-gather hooks (`_register_pre_backward_param_unshard_hook`) on every sub-module when `overlap_moe_expert_parallel_comm` + `optim_grads_params`, so backward all-gather works at the sub-module level.
- In per-parameter `register_post_accumulate_grad_hook`, skip parameters with `skip_backward_post_hook=True` so their gradient processing is deferred to the delayed wgrad path.

**`mcore_fsdp_adapter.py`** — Auto-configure FSDP for EP overlap
- Enable `enable_fine_grained_param_gather_hook` when `overlap_moe_expert_parallel_comm` is set (for all sharding strategies), so every sub-module gets its own forward all-gather/unshard hooks.
- Assert that `fsdp_double_buffer` and partial CUDA graph scopes are disabled when `overlap_moe_expert_parallel_comm` is set.
- Assert that `fsdp_unit_modules` must be `[TransformerLayer]` when combined with `optim_grads_params`.

**`model_chunk_schedule_plan.py`** — Per-layer FSDP hook wiring
- Add `set_fsdp_reshard_hooks()` to `TransformerLayerSchedulePlan` that attaches post-forward and post-backward release callbacks to the correct schedule nodes (last forward node, first backward node).

**`fine_grained_callables.py`** — Node-level FSDP hooks
- Override `forward()` / `backward()` in `TransformerLayerNode` to invoke FSDP reshard hooks at layer boundaries.
- In `backward_dw()`, collect and execute delayed `post_wgrad_grad_acc_hook` for FSDP gradient processing, then invoke the reshard hook.
- Add `set_fsdp_post_forward_reshard_hook()` / `set_fsdp_post_backward_reshard_hook()` registration methods.
- Add `parameters()` iterator to `_BackwardDWWrapper` so FSDP can discover parameters owned by the attn and shared-expert backward-dw callables.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by Wohox** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/4676377dc7f120c6dac96a3496a576f0b1e1d27f

**Comment by Wohox** [comment]
/ok to test [e5a27f0](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/e5a27f0818a5632a6cbe276b1e91bd3ffd9080c2)

**Comment by Wohox** [comment]
/ok to test [5c3847c](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/5c3847c5b495c30ee7ca67b7c9f3cf89dcc36f7d)

**Comment by Wohox** [comment]
/ok to test [1de0217](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/1de02175c722a417fbfe3fe33d62a10e961ec934)

**Comment by Wohox** [comment]
/ok to test [1a0c098](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/1a0c098b57f65ad38defc5ead30e808f154415e1)

**Comment by Wohox** [comment]
/ok to test [bbaca66](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/bbaca66188b2c896b21fe04b7d7dba52ac5a056b)

**Review by yaoyu-33 — APPROVED** [review]

File: tests/unit_tests/a2a_overlap/test_delay_wgrad_compute.py [hunk 1]
```diff
@@ -223,7 +179,95 @@ def test_overlap_dispatch_backward_with_experts_wgrad_with_fsdp(
...
+    @pytest.mark.parametrize("sharding_strategy", ["optim_grads_params", "optim_grads"])
+    def test_fsdp_1f1b_delay_wgrad(self, dispatcher_type, sharding_strategy):
+        """Verify FSDP + 1F1B overlap + delay_wgrad_compute.
+
+        Compares per-step loss and final weights between:
+          - Reference: FSDP + 1F1B overlap (without delay_wgrad_compute)
+          - Test:      FSDP + 1F1B overlap (with delay_wgrad_compute)
+
+        Both use combined_1f1b_schedule_for_no_pipelining with
+        overlap_moe_expert_parallel_comm.  The delay_wgrad_compute flag only
```
[the remark this answers is not in the document]
Comment (Wohox) [reply tests/unit_tests/a2a_overlap/test_delay_wgrad_compute.py:199]: added new test `test_fsdp_1f1b_overlap.py`

**Comment by Wohox** [comment]
/ok to test [a477539](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/a4775394361ee2dfd1fb9459fab10be5e3e676a8)

**Comment by Wohox** [comment]
/ok to test [cb83f70](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/cb83f70d4a99349f0820cb4f457f39b978305f44)

**Comment by Wohox** [comment]
/ok to test [db40f1d](https://github.com/NVIDIA/Megatron-LM/pull/3796/commits/db40f1d2ae957999c7934e946b46d00014fe15c9)
